### PR TITLE
[CALCITE-3929] When deserialize UDAF aggregate call from json string,…

### PR DESCRIPTION
**Pull request by xy2953396112**
… throws NPE 
https://issues.apache.org/jira/browse/CALCITE-3929

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java [hunk 1]
```diff
@@ -271,7 +271,9 @@ public RelDataType toType(RelDataTypeFactory typeFactory, Object o) {
 
   public Object toJson(AggregateCall node) {
     final Map<String, Object> map = jsonBuilder.map();
-    map.put("agg", toJson(node.getAggregation()));
+    Map<String, Object> aggMap = new HashMap(toJson(node.getAggregation()));
+    aggMap.put("class", node.getAggregation().getClass().getName());
```
Comment (michaelmior) [inline core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java:275]: Where is this `class` attribute used? It doesn't seem to be in [`RelJsonWriter#toAggCall`](https://github.com/apache/calcite/blob/e7d8d86dd3ab971921f5122de6c95868dec8739c/core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java#L278).

Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java:275]: Thank you for your comments. This is a problem of lack of class information during udaf serialization. When we deserialize udaf, we will try to load udaf function through class attribute[1].If we can't load SqlAggregateFunction correctly, there will be an exception when creating AggregateCall.
[[1]](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java#L680)

Comment (yanlin-Lynn) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java:275]: why new a map here, is it OK to do like this?
```
final Map<String, Object> aggMap = toJson(node.getAggregation());
aggMap.put("class", node.getAggregation().getClass().getName());
```
And you only need to add class info for UDAF, for system aggregate function, you do not need to add class info in the map.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java [hunk 2]
```diff
@@ -146,8 +147,9 @@
       + "        {\n"
       + "          \"agg\": {\n"
       + "            \"name\": \"COUNT\",\n"
-      + "            \"kind\": \"COUNT\",\n"
-      + "            \"syntax\": \"FUNCTION_STAR\"\n"
+      + "            \"syntax\": \"FUNCTION_STAR\",\n"
+      + "            \"class\": \"org.apache.calcite.sql.fun.SqlCountAggFunction\",\n"
+      + "            \"kind\": \"COUNT\"\n"
```
Comment (yanlin-Lynn) [inline core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:152]: I don't think class info is needed for system aggregate function

File: core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java [hunk 3]
```diff
@@ -125,6 +125,18 @@ public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
     }
   }
 
+  /** "MYAGGFUNC" user-defined aggregate function. This agg function accept one or more arguments
+   * in order to reproduce the throws of CALCITE-3929. */
+  public static class MyAggFunc extends SqlAggFunction {
```
Comment (yanlin-Lynn) [inline core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java:130]: extends `SqlUserDefinedAggFunction` might be better for your case

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java (hunk 1 above)
Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java]: Thanks for review.

File: core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java (hunk 3 above)
Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java]: There is a `MockSqlOperatorTable#MyAvgAggFunction`, we should use it to reproduce this case.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java (hunk 2 above)
Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java]: You are right. Code has been updated.

File: core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java (hunk 3 above)
Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/test/MockSqlOperatorTable.java:130]: @yanlin-Lynn
Thanks, `SqlUserDefinedAggFunction`  has an AggregateFunction argument, it's not convenient for reproduce this case.So it's better to extends SqlAggFunction and define it as a user-defined function.

**Comment by yanlin-Lynn** [comment]
LGTM

**Pull request merged**